### PR TITLE
ADD / TestSessionState initial implementation

### PR DESCRIPTION
**Pull request by dnsl48**
TestSessionState model initial implementation
TestSessionEnvironment to initialize the state for every scenario and provide API for the clients to use it
TestSessionHTTPMiddleware to keep the state fields up to date

### CODE REVIEW

**Comment by dnsl48** [comment]
Related to https://github.com/silverstripe/silverstripe-admin/issues/778

File: src/TestSessionState.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+
+/**
+ * The session state keeps some metadata about the current test session.
+ * This may allow the client (Behat) to get some insight into the
+ * server side affairs (e.g. if the server is handling some number requests at the moment).
+ *
+ * The client side (Behat) must not use this class straightforwardly, but rather
+ * rely on the API of {@see TestSessionEnvironment} or {@see TestSessionController}.
+ */
+class TestSessionState extends DataObject
```
[the remark this answers is not in the document]
Comment (dnsl48) [reply src/TestSessionState.php]: done

File: src/TestSessionHTTPMiddleware.php [hunk 2]
```diff
@@ -39,15 +41,42 @@ public function process(HTTPRequest $request, callable $delegate)
...
+            $this->decrementModelState();
         }
     }
 
+    protected function incrementModelState() {
+        $schema = DataObject::getSchema();
+
+        $update = SQLUpdate::create(sprintf('"%s"', $schema->tableName(TestSessionState::class)))
+            ->addWhere(['ID' => 1])
+            ->addAssignments([
```
[the remark this answers is not in the document]
Comment (dnsl48) [reply src/TestSessionHTTPMiddleware.php]: nice one, didn't know about this one, thanks :+1:

**Comment by maxime-rainville** [comment]
I meant to approve this, but forgot before the merge.

**Pull request merged**